circuit/primitives/src/xtx.rs

```diff
@@ -59,10 +65,458 @@ impl<
             delay_steps_at,
             result_status: None,
             total_reward,
+            full_side_effects,
         }
     }
 
     pub fn generate_xtx_id<T: frame_system::Config>(&self) -> XtxId<T> {
         SystemHashing::<T>::hash(Encode::encode(self).as_ref())
     }
+
+    pub fn is_completed(&self) -> bool {
+        for step in self.full_side_effects.iter() {
+            for full_side_effect in step.iter() {
+                if full_side_effect.confirmed.is_none() {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    // Complete the full side effect of Xtx by assigning confirmed side effect/
+    // This can only happen if the side effects is confirmed with respect to
+    // the execution steps.
+    //
+    // Return true if the input side effect successfully confirmed.
+    // Return false if no update has happened.
+    // Throw an error if detected the confirmation ruleset has been violated.
+    pub fn complete_side_effect<Hasher: sp_core::Hasher>(
+        &mut self,
+        confirmed: ConfirmedSideEffect<AccountId, BlockNumber, BalanceOf>,
+        input: SideEffect<AccountId, BlockNumber, BalanceOf>,
+    ) -> Result<bool, &'static str> {
+        let input_side_effect_id = input.generate_id::<Hasher>();
+
+        // Double check there are some side effects for that Xtx - should have been checked at API level tho already
+        if self.full_side_effects.is_empty() {
+            return Err("Xtx has no single side effect step to confirm");
+        }
+
+        let step_with_unconfirmed_steps: Option<usize>;
+
+        for (i, step) in self.full_side_effects.iter_mut().enumerate() {
+            // Double check there are some side effects for that Xtx - should have been checked at API level tho already
+            if step.is_empty() {
+                return Err("Xtx has an empty single step.");
+            }
+            for mut full_side_effect in step.iter_mut() {
+                if full_side_effect.confirmed.is_none() {
+                    step_with_unconfirmed_steps = Some(i);
+                    // Recalculate the ID for each input side effect and compare with the input one.
+                    // Check the current unconfirmed step before attempt to confirm the full side effect.
+                    return if full_side_effect.input.generate_id::<Hasher>() == input_side_effect_id
+                        && step_with_unconfirmed_steps == Some(i)
```

Comment:
i) We know that `step_with_unconfirmed_steps` equals `Some(i)` for sure, it's assigned above, do we still need to double-check that here?
ii) Do we need variable `step_with_unconfirmed_steps` at all?

Reply:
the inner loop here `Checks the current unconfirmed step before attempt to confirm the full side effect.`

var name `step_with_unconfirmed_steps` is actually confusing, would rather rename to `unconfirmed_step_no`

need to make sure that the side effects u are about to confirm is at the same `unconfirmed_step_no` and not in the future to respect the exec order